### PR TITLE
🌱 KCP: improve validation webhooks

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
While investigating #6113 I found out that our KCP validation has a few issues:
* field paths in KubeadmConfigSpec.Validate weren't correct in all cases (prefix)
* semantic version validation of CoreDNS image tag was only done on update, not on create
* errors in semantic version parsing weren't propagated in some cases
* KubeadmConfigSpec was not validated on KubeadmControlPlaneTemplate create

Overall I think that should make especially ClusterClass/KubeadmControlPlaneTemplate safer to use as the errors are already returned on creation of KubeadmControlPlaneTemplate. Not during topology reconcile when KubeadmControlPlane cannot be created.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #6113

### CODE REVIEW

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go [hunk 1]
```diff
@@ -107,7 +107,7 @@ func (c *KubeadmConfigSpec) validateFiles() field.ErrorList {
 				allErrs = append(
 					allErrs,
 					field.Invalid(
-						field.NewPath("spec", "files").Index(i).Child("contentFrom", "secret", "name"),
+						pathPrefix.Child("files").Index(i).Child("contentFrom", "secret", "name"),
```
Comment (killianmuldoon) [inline bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:110]: Not a blocker by any means, but what do you think about moving some of these to field.Required while we're making changes? The error seems more in line with what's being checked here ( and elsewhere for missing values)

File: controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go [hunk 2]
```diff
@@ -495,7 +508,7 @@ func (in *KubeadmControlPlane) validateCoreDNSVersion(prev *KubeadmControlPlane)
 			field.Invalid(
 				field.NewPath("spec", "kubeadmConfigSpec", "clusterConfiguration", "dns", "imageTag"),
```
Comment (killianmuldoon) [inline controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go:509]: Could we add the path prefix as a parameter in this function too?

**Review by killianmuldoon — CHANGES_REQUESTED** [review]
Overall looks good, but looks like it's a breaking change. We probably want to move these webhooks to an internal package like the others, right? Do you want to tackle that in this PR? Or we could revert the path change for this PR and include it in a change that makes the webhook implementation internal.

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go (hunk 1 above)
Comment (sbueringer) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:110]: Sounds good

File: controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go (hunk 2 above)
Comment (sbueringer) [reply controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go:509]: +1

**Comment by sbueringer** [comment]
> Overall looks good, but looks like it's a breaking change. We probably want to move these webhooks to an internal package like the others, right? Do you want to tackle that in this PR? Or we could revert the path change for this PR and include it in a change that makes the webhook implementation internal.

I'm not sure if we should apply the same stability guarantees to the Validate func as to our API usually. This func is not really meant to be used externally. It is just exported because KCP has to use it too.

I'm not sure if it's worth moving the webhook out of the API package as that is a breaking change itself (even though we could do that with deprecation).

Overall I think this breaking change should be acceptable as it improves error messages, probably nobody uses this func apart from us and it's very easy to adopt to this breaking change if folks upgrade to v1.2.0.

**Comment by killianmuldoon** [comment]
Would it be better to move them internally just to avoid someone asking the next time there's a PR. Seems like it could be  a good help wanted issue. :stuck_out_tongue_closed_eyes:

**Comment by sbueringer** [comment]
> Would it be better to move them internally just to avoid someone asking the next time there's a PR. Seems like it could be a good help wanted issue. 😝

I'm not sure. I think it has more downsides for us to move it elsewhere as it makes it just more complicated :)

But let's see what others think.

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go [hunk 3]
```diff
@@ -63,7 +63,7 @@ func (c *KubeadmConfig) ValidateDelete() error {
 }
 
 func (c *KubeadmConfigSpec) validate(name string) error {
-	allErrs := c.Validate()
+	allErrs := c.Validate(field.NewPath("spec"))
```
Comment (vincepri) [inline bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:66]: Is passing the spec field inside for code or is it solving an actual problem? While going through the changes, the older version with e.g. `"spec" "x" "y" "z"` is a bit more clear than the current proposed changes

Comment (sbueringer) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:66]: This solves the problem that we are reporting wrong field paths in our field errors in some webhooks.

We have different places where KubeadmConfigSpec are used, e.g. in KCP. But today all our webhooks report that the error is in `.spec...` without considering that it's actually e.g. `.spec.kubeadmConfigSpec...`

Furthermore (in my opinion) it's a best practice to build up those field paths incrementally, so each function only has to append new relative paths and we don't have to look up and down complex call hierarchies to figure out if the resulting field error has a correct field path. We are e.g. using that in our ClusterClass validation quite extensively (and in my opinion successfully :) ).

File: controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go (hunk 2 above)
Comment (sbueringer) [reply controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go:509]: Reconsidering. This function is called on the whole KubeadmControlPlane (not on a sub-struct). It is also only used for KubeadmControlPlane (KubeadmControlPlaneTemplates are immutable).

I would keep it as is.

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go (hunk 1 above)
Comment (killianmuldoon) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:110]: Sounds  good to me - I'll do a bit of a scan and open an issue around it.

Comment (sbueringer) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:110]: Perfect, thank you!

**Comment by sbueringer** [comment]
@vincepri PTAL, if you have some time, thank you! :)

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go (hunk 1 above)
Comment (killianmuldoon) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:110]: #6249 covers this I think.

Comment (sbueringer) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go:110]: Thx!

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**